File: filesync/src/main/java/de/mel/filesync/bash/AutoKlausIterator.java
package de.mel.filesync.bash;

import java.util.Iterator;

public interface AutoKlausIterator<T> extends Iterator<T>, AutoCloseable {

    public static class EmpyAutoKlausIterator<T> implements AutoKlausIterator<T>{
        @Override
        public void close() throws Exception {

        }

        @Override
        public boolean hasNext() {
            return false;
        }

        @Override
        public T next() {
            return null;
        }
    }
}

File: filesync/src/main/java/de/mel/filesync/bash/AutoKlausIterator.kt
package de.mel.filesync.bash

interface AutoKlausIterator<out T> : Iterator<T>, AutoCloseable {
    class EmpyAutoKlausIterator<T> : AutoKlausIterator<T> {
        @Throws(Exception::class)
        override fun close() {
        }

        override fun hasNext(): Boolean {
            return false
        }

        override fun next(): T = throw IllegalAccessException("no element left")
    }

    class EmptyAutoKlausIterator<out O> : AutoKlausIterator<O>{
        override fun hasNext(): Boolean  = false

        override fun next(): O = throw IllegalAccessException("no element left")

        override fun close() {
        }
    }
}